### PR TITLE
[RFC] Remove project-specific integer types: long_u. (5)

### PR DESCRIPTION
**Pull request by elmart**
Continues #1788.

Removes all remaining `long_u` instances. Finally!

### CODE REVIEW

File: src/nvim/ex_docmd.c [hunk 1]
```diff
@@ -2204,7 +2205,7 @@ find_ucmd (
             eap->cmdidx = CMD_USER;
           else
             eap->cmdidx = CMD_USER_BUF;
-          eap->argt = (long)uc->uc_argt;
+          eap->argt = uc->uc_argt;
```
Comment (oni-link) [inline src/nvim/ex_docmd.c]: Member `argt` of  `struct exarg` has type `long`. Are we using additional infos why the value cannot be bigger than `INT_MAX`?

Comment (elmart) [inline src/nvim/ex_docmd.c]: Note that here we are assigning an `uint32_t` to a `long`, not the other way. So, there's no problem.
The variable being assigned to (eap->argt) is of type long.
The value being assigned (uc->uc_argt) is a member of struct ucmd, of type uint32_t. I
Before refactoring (struct ucmd).uc_argt, I checked all values assigned to it never needed more than 4 bytes, and were always positive. Thus I refactored it into an uint32_t. Probably, the same will be true for (struct exarg).argt, but I haven't checked that yet.
This was one of many places were long (and long_u) are used when only 4 bytes are required (which had sense for 16bit platforms, but not for 32bit and above).

What it's true, it's that there's a missing cast here, to avoid warning for mixing signed/unsigned types, which is not being signaled because -Wconversion is not active on ex_docmd.c. The line should be:
```
eap->argt = (long)uc->uc_argt
```

Comment (elmart) [inline src/nvim/ex_docmd.c]: That last statement is not exactly correct. Let me rephrase:
For 64bit, uint32_t is always safely convertible to long, so everything's correct, even without cast.
For 32bit, warning would be generated, for assigning unsigned to signed of the same size.
Now, we know unsigned value never uses the last bit, so a simple cast to long would suffice.
But that would be fragile, as it would break if new values using the last bit were introduced.
So, proper solution would be to assert assigned value is not bigger than LONG_MAX.
But that would generate and "always true" warning in 64 bits.
So, definitive thing should look like:
```
assert((uintmax_t)uc->uc_argt <= LONG_MAX);
eap->argt = (long)uc->uc_argt
```

I will correct to that if you agree.

Comment (elmart) [inline src/nvim/ex_docmd.c]: Note that this all will disappear once eap->argt is refactored into uint32_t too (if that's possible, as I presume).

Comment (elmart) [inline src/nvim/ex_docmd.c]: Bad luck. 
```
assert((uintmax_t)uc->uc_argt <= LONG_MAX);
```
still throws the "always true" warning on 64bits.
I seem not to be able to find an expression that doesn't generate a warning in both 32 and 64bits.
Only solution I have by now is using preprocessor:
```
#if SIZEOF_LONG <= 4
    assert(uc->uc_argt <= LONG_MAX);
#endif
```

I haven't needed that until now. Can you think of other option?

Comment (oni-link) [inline src/nvim/ex_docmd.c]: I used `grep 'argt\>' src/nvim/* -r` to look for `(struct exarg).argt`. There are only a few places where an assignment is used (`=`,`|=`). Only `uint32_t` values are assigned, so I think it would be easier to just change the type from `long` to `uint32_t` for `(struct exarg).argt`.

Comment (elmart) [inline src/nvim/ex_docmd.c]: Ok. I'll try that just now. I'll add a new commit (not repush) with the correction when finished, so that you can continue reviewing knowing nothing will change under your feet. I will reintegrate those review commits to their corresponding place when review has finished.

Comment (elmart) [inline src/nvim/ex_docmd.c]: Done.

File: src/nvim/misc1.c [hunk 2]
```diff
@@ -1292,11 +1293,12 @@ int plines_win_nofold(win_T *wp, linenr_T lnum)
   width = wp->w_width - win_col_off(wp);
   if (width <= 0)
     return 32000;
-  if (col <= width)
+  if (col <= (unsigned int)width)
     return 1;
   col -= width;
   width += win_col_off2(wp);
-  return (col + (width - 1)) / width + 1;
+  assert(col <= INT_MAX);
```
Comment (oni-link) [inline src/nvim/misc1.c]: A line should not be longer than `INT32_MAX` (`:h limits`, `win_linetabsize()`) and after a possible `col += 1` and `col -= width`, we should use `assert((int)col < INT32_MAX - (width - 1));`.

File: src/nvim/option.c [hunk 3]
```diff
@@ -6497,8 +6499,10 @@ void comp_col(void)
     if (!p_ru || last_has_status)           /* no need for separating space */
       ++sc_col;
   }
-  sc_col = Columns - sc_col;
-  ru_col = Columns - ru_col;
+  assert((long)INT_MIN + sc_col <= Columns && Columns <= (long)INT_MAX + sc_col);
```
Comment (oni-link) [inline src/nvim/option.c]: For `LONG_MAX` equal `INT_MAX` this won't work: `(long) INT_MAX + sc_col`

File: src/nvim/option.c [hunk 4]
```diff
@@ -6497,8 +6499,10 @@ void comp_col(void)
...
+  sc_col = (int)(Columns - sc_col);
+  assert((long)INT_MIN + ru_col <= Columns && Columns <= (long)INT_MAX + ru_col);
```
Comment (oni-link) [inline src/nvim/option.c]: For `LONG_MAX` equal `INT_MAX` this won't work: `(long) INT_MAX + ru_col`

File: src/nvim/misc1.c (hunk 2 above)
Comment (elmart) [inline src/nvim/misc1.c]: Here I was just "mechanically" asserting following cast is safe.
I'm not sure if this is the place to also check for variable "conceptual" limits. We're not changing a persistent variable holding a column number, but just making a calculation of a different thing out of it.
Also, we're adding at most 1, and subtracting at least 1. So, if value returned by by win_linetabsize() is in bounds, so will be the newly calculated value.
I don't have any problem to add the assert if you prefer, though.

Comment (oni-link) [inline src/nvim/misc1.c]: If the line length could really reach `INT32_MAX` and `col += 1` is used, we could have an overlfow for a not changing `width` of `1`:
`(INT_MAX+(1-1))/1+1 > INT_MAX`.

File: src/nvim/option.c (hunk 3 above)
Comment (elmart) [inline src/nvim/option.c]: What about this?
```
assert(INT_MIN + sc_col <= Columns && Columns - sc_col <= INT_MAX);
```

File: src/nvim/option.c (hunk 4 above)
Comment (elmart) [inline src/nvim/option.c]: Same as above.

File: src/nvim/option.c (hunk 3 above)
Comment (oni-link) [inline src/nvim/option.c]: :+1:

File: src/nvim/misc1.c (hunk 2 above)
Comment (elmart) [inline src/nvim/misc1.c]: Note that `col` is now `unsigned int`. You can safely add 1, even if it equals INT32_MAX.

Comment (oni-link) [inline src/nvim/misc1.c]: Yes, but my problem is in the return statement. Everything there is an `int`, so the expression should also be evaluated as an `int`.

File: src/nvim/option.c (hunk 3 above)
Comment (elmart) [inline src/nvim/option.c]: Added review commit.

File: src/nvim/option.c [hunk 5]
```diff
@@ -7867,20 +7872,18 @@ opt_strings_flags (
...
       if (values[i] == NULL)            /* val not found in values[] */
         return FAIL;
 
-      len = (int)STRLEN(values[i]);
+      size_t len = STRLEN(values[i]);
       if (STRNCMP(values[i], val, len) == 0
           && ((list && val[len] == ',') || val[len] == NUL)) {
         val += len + (val[len] == ',');
-        new_flags |= (1 << i);
+        new_flags |= (1U << i);
```
Comment (oni-link) [inline src/nvim/option.c]: Do we need to check the shift width or was this always well defined for shifting an unsigned value?

File: src/nvim/misc1.c (hunk 2 above)
Comment (elmart) [inline src/nvim/misc1.c]: Ah, ok. Now I get you. 
A couple of gotchas with the proposed assert:
- `(int)col` would be unsafe without first checking col < INT_MAX.
- We are doing calculations as `int`, not `int32_t`. So, it should be INT_MAX, not INT32_MAX.

So, it would remain:
```
assert(col <= INT_MAX && (int)col < INT_MAX - (width - 1));
```
Agree?

Comment (oni-link) [inline src/nvim/misc1.c]: :+1:

File: src/nvim/option.c (hunk 5 above)
Comment (elmart) [inline src/nvim/option.c]: Didn't check for that.
But note I changed `i` from `int` to `unsigned int` (same size).
So, if it worked before, it should work now, shouldn't it?

File: src/nvim/misc1.c (hunk 2 above)
Comment (elmart) [inline src/nvim/misc1.c]: Added review commit.

File: src/nvim/option.c (hunk 5 above)
Comment (oni-link) [inline src/nvim/option.c]: My edge case scenario: Someone calls this function with at least `sizeof(unsigned)*8` entries in `values`. So it could happen that we try to shift more bits than the first operand has (my assumption here is that this is undefined).

**Comment by elmart** [comment]
Sorry, I had to repush because build was failing for reasons other than this PR.

File: src/nvim/option.c (hunk 5 above)
Comment (elmart) [inline src/nvim/option.c]: Longest array with which is called seems to be `p_ssop_values`, which has 15 values.

Comment (aktau) [inline src/nvim/option.c]: We could either add an assert to ensure that this variable shift stays within bounds, or we could trust in (and us) the ubsan (undefined behaviour sanitizer).

Comment (elmart) [inline src/nvim/option.c]: I'll add the assert. Not sure if ubsan could pick such a thing.

Comment (elmart) [inline src/nvim/option.c]: Added review commit.

File: src/nvim/option.c [hunk 6]
```diff
@@ -4667,7 +4664,9 @@ char_u *check_colorcolumn(win_T *wp)
       col = col * getdigits_int(&s);
       if (wp->w_buffer->b_p_tw == 0)
         goto skip;          /* 'textwidth' not set, skip this item */
-      col += wp->w_buffer->b_p_tw;
+      assert(wp->w_buffer->b_p_tw >= (long)INT_MIN - col
+             && wp->w_buffer->b_p_tw <= (long)INT_MAX - col);
```
Comment (elmart) [inline src/nvim/option.c]: This won't work when sizeof(long) == sizeof(int).
Following the same spirit as before, I could change it to:
```
       assert((col >= 0                                                           
               && wp->w_buffer->b_p_tw <= INT_MAX - col                           
               && wp->w_buffer->b_p_tw + col >= INT_MIN)                          
              || (col < 0                                                         
                  && wp->w_buffer->b_p_tw >= INT_MIN - col                        
                  && wp->w_buffer->b_p_tw + col <= INT_MAX));
```
But I've found it's easier if we test the opposite (not overflow condition instead of in-bounds condition):
```
      assert(!(col > 0 && wp->w_buffer->b_p_tw > INT_MAX - col                   
                || col < 0 && wp->w_buffer->b_p_tw < INT_MIN - col));             
```
I will review other asserts, to follow this shorter (and IMO, clearer) style.

Comment (oni-link) [inline src/nvim/option.c]: `col == 0`  also needs to be checked: `col >= 0 && ... || col <= 0 && ...`

Comment (elmart) [inline src/nvim/option.c]: Arrrgh. Right :+1: Former expression would be valid if all terms were int. But not when we're checking for narrowing type conversion. Your precision continues to marvel me, @oni-link ;-)

Comment (elmart) [inline src/nvim/option.c]: Added review commit with some fixed asserts in option.c.
I think there are other fixable asserts in other files, but can't continue working just now. I'll resume later this evening.

**Comment by elmart** [comment]
@oni-link Added a couple of review commits fixing asserts in option.c and regexp-nfa.c.
You can continue reviewing when you want.

**Comment by elmart** [comment]
These kind of asserts are tricky. Even when you have done before, it's easy to make mistakes. So, I'm thinking about the introduction of some macros to check aritmetic operations do not overflow.
Something like: `SIGNED_SUM_IN_BOUNDS(operand_1, operand_2, type)`, that would expand to
```
(!((operand_2 >= 0 && operand_1 > type_MAX - operand_2)
    || (operand_2 <= 0 && operand_1 < type_MIN - operand_2))))
```
to be used like
```
assert(<possibly other conditions>...
       && SIGNED_SUM_IN_BOUNDS(val1, val2, long));
```
Same for SIGNED_SUBSTRACTION_IN_BOUNDS, UNSIGNED_SUM_IN_BOUNDS, etc...

That would make much easier for people to insert the correct checkings wherever necessary. And would be much easier to review, too. What do you think?

**Comment by oni-link** [comment]
> That would make much easier for people to insert the correct checkings wherever necessary. And would be much easier to review, too. What do you think?

That would be great.

> @oni-link Added a couple of review commits fixing asserts in option.c and regexp-nfa.c.
You can continue reviewing when you want.

If you replace some of the asserts with these new macros I will wait with reviewing the last commits.

**Comment by elmart** [comment]
I'd prefer leaving the macros for a separated PR. I have to think a bit about the optimal shape for them. Also, that way I could review asserts introduced previously, and convert all the suitable ones to use the macros.

**Comment by oni-link** [comment]
LGTM.

**Comment by elmart** [comment]
Review commits integrated, and PR rebased.
Thanks so much, @oni-link, for your immense help with those tricky asserts.

**Comment by justinmk** [comment]
Thank you both. Amazing work.

**Comment by aktau** [comment]
Props!
